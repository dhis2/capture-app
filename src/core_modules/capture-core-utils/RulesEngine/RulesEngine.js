// @flow
import VariableService from './VariableService/VariableService';
import ValueProcessor from './ValueProcessor/ValueProcessor';
import getExecutionService from './executionService/executionService';
import getDateUtils from './dateUtils/dateUtils';
import getRulesEffectsProcessor from './rulesEffectsProcessor/rulesEffectsProcessor';
import processTypes from './rulesEffectsProcessor/processTypes.const';

import type {
    OutputEffect,
    IConvertInputRulesValue,
    IConvertOutputRulesEffectsValue,
    IMomentConverter,
    ProgramRulesContainer,
    EventData,
    EventsData,
    DataElements,
    OrgUnit,
    OptionSets,
    Translator,
    TrackedEntityAttributes,
    Enrollment,
    ProgramRuleEffect,
    EventsDataContainer,
    TEIValues,
} from './rulesEngine.types';


type ExecutionService = {
    executeRules: (
        programRulesContainer: ProgramRulesContainer,
        executingEvent: ?EventData | {},
        events: ?EventsDataContainer,
        dataElements: ?DataElements,
        trackedEntityAttributes: ?TrackedEntityAttributes,
        selectedTrackedEntityAttributes: ?TEIValues,
        selectedEnrollment: ?Enrollment,
        selectedOrgUnit: OrgUnit,
        optionSets: ?OptionSets,
        flags: Object,
    ) => ?Array<ProgramRuleEffect>,
    convertDataToBaseOutputValue: (data: any, valueType: string) => any,
};

export default class RulesEngine {
    executionService: ExecutionService;
    onProcessRulesEffects: (
        effects: Array<ProgramRuleEffect>,
        processType: $Values<typeof processTypes>,
        dataElements: ?DataElements,
        trackedEntityAttributes?: ?TrackedEntityAttributes) => ?Array<OutputEffect>;

    constructor(
        inputConverterObject: IConvertInputRulesValue,
        momentConverter: IMomentConverter,
        onTranslate: Translator,
        outputRulesConverterObject: IConvertOutputRulesEffectsValue) {
        const valueProcessor = new ValueProcessor(inputConverterObject);
        const dateUtils = getDateUtils(momentConverter);
        const variableService = new VariableService(valueProcessor.processValue, dateUtils);
        this.executionService = getExecutionService(onTranslate, variableService, dateUtils);
        this.onProcessRulesEffects = getRulesEffectsProcessor(this.executionService.convertDataToBaseOutputValue, outputRulesConverterObject);
    }

    executeRules(
        programRulesContainer: ProgramRulesContainer,
        executingEvent: ?EventData | {},
        eventsData: ?EventsData,
        dataElements: ?DataElements,
        enrollmentData: ?Enrollment,
        teiValues: ?TEIValues,
        trackedEntityAttributes: ?TrackedEntityAttributes,
        selectedOrgUnit: OrgUnit,
        optionSets: ?OptionSets,
        processType: $Values<typeof processTypes>,
    ): ?Array<OutputEffect> {
        let eventsContainer;
        if (eventsData && eventsData.length > 0) {
            const eventsDataByStage = eventsData.reduce((accEventsByStage, event) => {
                const hasProgramStage = !!event.programStageId;
                if (hasProgramStage) {
                    accEventsByStage[event.programStageId] = accEventsByStage[event.programStageId] || [];
                    accEventsByStage[event.programStageId].push(event);
                }
                return accEventsByStage;
            }, {});

            eventsContainer = {
                all: eventsData,
                byStage: eventsDataByStage,
            };
        } else {
            eventsContainer = null;
        }

        const effects = this.executionService.executeRules(
            programRulesContainer,
            executingEvent,
            eventsContainer,
            dataElements,
            trackedEntityAttributes,
            teiValues,
            enrollmentData,
            selectedOrgUnit,
            optionSets,
            { debug: true },
        );

        const processedEffects = effects ?
            this.onProcessRulesEffects(effects, processType, dataElements, trackedEntityAttributes) :
            null;
        return processedEffects;
    }
}
