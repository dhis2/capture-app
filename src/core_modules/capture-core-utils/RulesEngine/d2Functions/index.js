import log from 'loglevel';
import getZScoreWFA from './zScoreWFA';

export default function d2Functions(dateUtils, variableService, variablesHash) {
    return {
        'd2:concatenate': {
            name: 'd2:concatenate',
            dhisFunction: (params) => {
                let returnString = "'";
                for (let i = 0; i < params.length; i++) {
                    returnString += params[i];
                }
                returnString += "'";
                return returnString;
            },
        },
        'd2:daysBetween': {
            name: 'd2:daysBetween',
            parameters: 2,
            dhisFunction: params => dateUtils.daysBetween(params[0], params[1]),
        },
        'd2:weeksBetween': {
            name: 'd2:weeksBetween',
            parameters: 2,
            dhisFunction: params => dateUtils.weeksBetween(params[0], params[1]),
        },
        'd2:monthsBetween': {
            name: 'd2:monthsBetween',
            parameters: 2,
            dhisFunction: params => dateUtils.monthsBetween(params[0], params[1]),
        },
        'd2:yearsBetween': {
            name: 'd2:yearsBetween',
            parameters: 2,
            dhisFunction: params => dateUtils.yearsBetween(params[0], params[1]),
        },
        'd2:floor': {
            name: 'd2:floor',
            parameters: 1,
            dhisFunction: params => Math.floor(params[0]),
        },
        'd2:modulus': {
            name: 'd2:modulus',
            parameters: 2,
            dhisFunction: (params) => {
                const dividend = Number(params[0]);
                const divisor = Number(params[1]);
                const rest = dividend % divisor;
                return rest;
            },
        },
        'd2:addDays': {
            name: 'd2:addDays',
            parameters: 2,
            dhisFunction: (params) => {
                const date = params[0];
                const daysToAdd = params[1];
                return dateUtils.addDays(date, daysToAdd);
            },
        },
        'd2:zing': {
            name: 'd2:zing',
            parameters: 1,
            dhisFunction: (params) => {
                let number = params[0];
                if (number < 0) {
                    number = 0;
                }
                return number;
            },

        },
        'd2:oizp': {
            name: 'd2:oizp',
            parameters: 1,
            dhisFunction: (params) => {
                const number = params[0];
                let output = 1;
                if (number < 0) {
                    output = 0;
                }
                return output;
            },
        },
        'd2:count': {
            name: 'd2:count',
            parameters: 1,
            dhisFunction: (params) => {
                const variableName = params[0];
                const variableObject = variablesHash[variableName];
                let count = 0;
                if (variableObject) {
                    if (variableObject.hasValue) {
                        if (variableObject.allValues) {
                            count = variableObject.allValues.length;
                        } else {
                            // If there is a value found for the variable, the count is 1 even if there is no list of alternate values
                            // This happens for variables of "DATAELEMENT_CURRENT_STAGE" and "TEI_ATTRIBUTE"
                            count = 1;
                        }
                    }
                } else {
                    log.warn(`could not find variable to count: ${variableName}`);
                }
                return count;
            },
        },
        'd2:countIfZeroPos': {
            name: 'd2:countIfZeroPos',
            parameters: 1,
            dhisFunction: (params) => {},
        },
        'd2:countIfValue': {
            name: 'd2:countIfValue',
            parameters: 2,
            dhisFunction: (params) => {
                const variableName = params[0];
                const variableObject = variablesHash[variableName];
                const valueToCompare = variableService.processValue(params[1], variableObject.variableType);

                let count = 0;
                if (variableObject) {
                    if (variableObject.hasValue) {
                        if (variableObject.allValues) {
                            for (let i = 0; i < variableObject.allValues.length; i++) {
                                if (valueToCompare === variableObject.allValues[i]) {
                                    count += 1;
                                }
                            }
                        } else if (valueToCompare === variableObject.variableValue) {
                            // The variable has a value, but no list of alternates. This means we compare the standard variablevalue
                            count = 1;
                        }
                    }
                } else {
                    log.warn(`could not find variable to countifvalue: ${variableName}`);
                }
                return count;
            },
        },
        'd2:ceil': {
            name: 'd2:ceil',
            parameters: 1,
            dhisFunction: params => Math.ceil(params[0]),
        },
        'd2:round': {
            name: 'd2:round',
            parameters: 1,
            dhisFunction: params => Math.round(params[0]),
        },
        'd2:hasValue': {
            name: 'd2:hasValue',
            parameters: 1,
            dhisFunction: (params) => {
                let valueFound = false;
                // when value of a field is equal to 0 we want the function to return true
                if (params[0] || params[0] === 0) {
                    valueFound = true;
                }
                return valueFound;
            },
        },
        'd2:lastEventDate': {
            name: 'd2:lastEventDate',
            parameters: 1,
            dhisFunction: (params) => {
                const variableName = params[0];
                const variableObject = variablesHash[variableName];
                let valueFound = "''";
                if (variableObject) {
                    if (variableObject.variableEventDate) {
                        valueFound = variableService.processValue(variableObject.variableEventDate, 'DATE');
                    } else {
                        log.warn(`no last event date found for variable: ${variableName}`);
                    }
                } else {
                    log.warn(`could not find variable to check last event date: ${variableName}`);
                }
                return valueFound;
            },
        },
        'd2:validatePattern': {
            name: 'd2:validatePattern',
            parameters: 2,
            dhisFunction: (params) => {
                const inputToValidate = params[0].toString();
                const pattern = params[1];
                const regEx = new RegExp(pattern, 'g');
                const match = inputToValidate.match(regEx);

                let matchFound = false;
                if (match !== null && inputToValidate === match[0]) {
                    matchFound = true;
                }
                return matchFound;
            },
        },
        'd2:addControlDigits': {
            name: 'd2:addControlDigits',
            parameters: 1,
            dhisFunction: (params) => {
                const baseNumber = params[0];
                const baseDigits = baseNumber.split('');
                const error = false;

                let firstDigit = 0;
                let secondDigit = 0;

                if (baseDigits && baseDigits.length < 10) {
                    let firstSum = 0;
                    const baseNumberLength = baseDigits.length;
                    // weights support up to 9 base digits:
                    const firstWeights = [3, 7, 6, 1, 8, 9, 4, 5, 2];
                    for (let i = 0; i < baseNumberLength && !error; i++) {
                        firstSum += parseInt(baseDigits[i], 10) * firstWeights[i];
                    }
                    firstDigit = firstSum % 11;

                    // Push the first digit to the array before continuing, as the second digit is a result of the
                    // base digits and the first control digit.
                    baseDigits.push(firstDigit);
                    // Weights support up to 9 base digits plus first control digit:
                    const secondWeights = [5, 4, 3, 2, 7, 6, 5, 4, 3, 2];
                    let secondSum = 0;
                    for (let si = 0; si < baseNumberLength + 1 && !error; si++) {
                        secondSum += parseInt(baseDigits[si], 10) * secondWeights[si];
                    }
                    secondDigit = secondSum % 11;

                    if (firstDigit === 10) {
                        log.warn('First control digit became 10, replacing with 0');
                        firstDigit = 0;
                    }
                    if (secondDigit === 10) {
                        log.warn('Second control digit became 10, replacing with 0');
                        secondDigit = 0;
                    }
                } else {
                    log.warn(`Base number not well formed(${baseDigits.length} digits): ${baseNumber}`);
                }

                if (!error) {
                    // Replace the end evaluation of the dhis function:
                    return baseNumber + firstDigit + secondDigit;
                }
                // Replace the end evaluation of the dhis function:
                return baseNumber;
            },
        },
        'd2:checkControlDigits': {
            name: 'd2:checkControlDigits',
            parameters: 1,
            dhisFunction: (params) => {
                log.warn('checkControlDigits not implemented yet');
                return params[0];
            },
        },
        'd2:left': {
            name: 'd2:left',
            parameters: 2,
            dhisFunction: (params) => {
                const string = String(params[0]);
                const numChars = string.length < params[1] ? string.length : params[1];
                const returnString = string.substring(0, numChars);
                return variableService.processValue(returnString, 'TEXT');
            },
        },
        'd2:right': {
            name: 'd2:right',
            parameters: 2,
            dhisFunction: (params) => {
                const string = String(params[0]);
                const numChars = string.length < params[1] ? string.length : params[1];
                const returnString = string.substring(string.length - numChars, string.length);
                return variableService.processValue(returnString, 'TEXT');
            },
        },
        'd2:substring': {
            name: 'd2:substring',
            parameters: 3,
            dhisFunction: (params) => {
                const string = String(params[0]);
                const startChar = string.length < params[1] - 1 ? -1 : params[1];
                const endChar = string.length < params[2] ? -1 : params[2];
                if (startChar < 0 || endChar < 0) {
                    return "''";
                }
                const returnString = string.substring(startChar, endChar);
                return variableService.processValue(returnString, 'TEXT');
            },
        },
        'd2:split': {
            name: 'd2:split',
            parameters: 3,
            dhisFunction: (params) => {
                const string = String(params[0]);
                const splitArray = string.split(params[1]);
                let returnPart = '';
                if (splitArray.length >= params[2]) {
                    returnPart = splitArray[params[2]];
                }
                return variableService.processValue(returnPart, 'TEXT');
            },
        },
        'd2:zScoreWFA': {
            name: 'd2:zScoreWFA',
            parameters: 3,
            dhisFunction: params => getZScoreWFA(params[0], params[1], params[2]),
        },
        'd2:length': {
            name: 'd2:length',
            parameters: 1,
            dhisFunction: params => String(params[0]).length,

        },
    };
}
