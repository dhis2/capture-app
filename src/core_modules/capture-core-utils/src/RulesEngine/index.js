// @flow
export { default as RulesEngine } from './RulesEngine';
export { default as effectActions } from './effectActions.const';
export { default as processTypes } from './rulesEffectsProcessor/processTypes.const';
