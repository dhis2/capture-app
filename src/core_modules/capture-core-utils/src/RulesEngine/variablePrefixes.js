// @flow

export default {
    DATAELEMENT: '#',
    TRACKED_ENTITY_ATTRIBUTE: 'A',
    CONTEXT_VARIABLE: 'V',
    CONSTANT_VARIABLE: 'C',
    CALCULATED_VALUE: '#',
};
