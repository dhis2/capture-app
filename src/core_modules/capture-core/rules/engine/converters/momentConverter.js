// @flow
import moment from 'moment';
import { getFormattedStringFromMomentUsingEuropeanGlyphs } from '../../../../capture-core-utils/date';
import type { IMomentConverter } from '../rulesEngine.types';

const momentFormat = 'YYYY-MM-DD';

class RulesMomentConverter implements IMomentConverter {
    // eslint-disable-next-line class-methods-use-this
    rulesDateToMoment(rulesEngineValue: string): moment$Moment {
        return moment(rulesEngineValue, momentFormat);
    }
    // eslint-disable-next-line class-methods-use-this
    momentToRulesDate(momentObject: moment$Moment): string {
        return getFormattedStringFromMomentUsingEuropeanGlyphs(momentObject);
    }
}

export const momentConverter = new RulesMomentConverter();
