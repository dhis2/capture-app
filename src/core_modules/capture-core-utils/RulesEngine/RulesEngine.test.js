// @flow
import i18n from '@dhis2/d2-i18n';

import RulesEngine from '../RulesEngine/RulesEngine';
import inputValueConverter from '../../capture-core/rulesEngineActionsCreator/converters/inputValueConverter';
import momentConverter from '../../capture-core/rulesEngineActionsCreator/converters/momentConverter';
import outputRulesEffectsValueConverter from '../../capture-core/rulesEngineActionsCreator/converters/rulesEffectsValueConverter';
import runRulesForSingleEvent from '../../capture-core/rulesEngineActionsCreator/runRulesForSingleEvent';

const programs = [
    {
        program: {
            programRules: [{ id: 'GC4gpdoSD4r', condition: '#{hemoglobin} < 9', description: 'Show warning if hemoglobin is dangerously low', displayName: 'Hemoglobin warning', programId: 'lxAQ7Zs9VYR', programRuleActions: [{ id: 'suS9GnraCx1', content: 'Hemoglobin value lower than normal', dataElementId: 'vANAXwtLwcT', programRuleActionType: 'SHOWWARNING' }] }, { id: 'dahuKlP7jR2', condition: '#{hemoglobin} > 99', description: 'Show error for hemoglobin value higher than 99', displayName: 'Show error for high hemoglobin value', programId: 'lxAQ7Zs9VYR', programRuleActions: [{ id: 'UUwZWS8uirn', content: 'The hemoglobin value cannot be above 99', dataElementId: 'vANAXwtLwcT', programRuleActionType: 'SHOWERROR' }] }, { id: 'hk30qiUJYUR', condition: '#{Christos } == true', displayName: 'Christos Rules', programId: 'lxAQ7Zs9VYR', programRuleActions: [{ id: 'eotNEY9CWxU', content: 'SAY YES', dataElementId: 'Ok9OQpitjQr', programRuleActionType: 'SHOWERROR' }] }, { id: 'xOe5qCzRS0Y', condition: '!#{womanSmoking} ', description: 'Hide smoking cessation councelling dataelement unless patient is smoking', displayName: 'Hide smoking cessation councelling', programId: 'lxAQ7Zs9VYR', programRuleActions: [{ id: 'hwgyO59SSxu', dataElementId: 'Ok9OQpitjQr', programRuleActionType: 'HIDEFIELD' }] }],
            programRuleVariables: [{ id: 'DINatbKMS71', dataElementId: 'Ok9OQpitjQr', displayName: 'Christos ', programId: 'lxAQ7Zs9VYR', programRuleVariableSourceType: 'DATAELEMENT_CURRENT_EVENT', useNameForOptionSet: true }, { id: 'Z92dJO9gIje', dataElementId: 'sWoqcoByYmD', displayName: 'womanSmoking', programId: 'lxAQ7Zs9VYR', programRuleVariableSourceType: 'DATAELEMENT_NEWEST_EVENT_PROGRAM', useNameForOptionSet: true }, { id: 'omrL0gtPpDL', dataElementId: 'vANAXwtLwcT', displayName: 'hemoglobin', programId: 'lxAQ7Zs9VYR', programRuleVariableSourceType: 'DATAELEMENT_NEWEST_EVENT_PROGRAM', useNameForOptionSet: true }],
            id: 'lxAQ7Zs9VYR',
        },
        foundation: { programRules: [] },
        orgUnit: { id: 'DiszpKrYNg8', name: 'Ngelehun CHC' },
    },
    {
        program: {
            programRules: [{ id: 'fd3wL1quxGb', condition: "#{gender} == 'Male'", description: 'Hide pregnant if gender is male', displayName: 'Hide pregnant if gender is male', programId: 'eBAyeGv0exc', programRuleActions: [{ id: 'IrmpncBsypT', dataElementId: 'SWfdB5lX0fk', programRuleActionType: 'HIDEFIELD', programStageSectionId: 'd7ZILSbPgYh' }] }, { id: 'x7PaHGvgWY2', condition: 'true', description: 'Body Mass Index. Weight in kg / height in m square.', displayName: 'BMI', programId: 'eBAyeGv0exc', programRuleActions: [{ id: 'x7PaHGvgWY2', content: 'BMI', data: '#{Zj7UnCAulEk.vV9UWAZohSf}/((#{Zj7UnCAulEk.GieVkTxp4HH}/100)*(#{Zj7UnCAulEk.GieVkTxp4HH}/100))', programRuleActionType: 'DISPLAYKEYVALUEPAIR', location: 'indicators' }] }],
            programRuleVariables: [{ id: 'RycV5uDi66i', dataElementId: 'qrur9Dvnyt5', displayName: 'age', programId: 'eBAyeGv0exc', programRuleVariableSourceType: 'DATAELEMENT_NEWEST_EVENT_PROGRAM', useNameForOptionSet: true }, { id: 'zINGRka3g9N', dataElementId: 'oZg33kd9taw', displayName: 'gender', programId: 'eBAyeGv0exc', programRuleVariableSourceType: 'DATAELEMENT_NEWEST_EVENT_PROGRAM', useNameForOptionSet: true }, { id: 'Zj7UnCAulEk.vV9UWAZohSf', displayName: 'Zj7UnCAulEk.vV9UWAZohSf', programRuleVariableSourceType: 'DATAELEMENT_CURRENT_EVENT', dataElementId: 'vV9UWAZohSf', programId: 'eBAyeGv0exc' }, { id: 'Zj7UnCAulEk.GieVkTxp4HH', displayName: 'Zj7UnCAulEk.GieVkTxp4HH', programRuleVariableSourceType: 'DATAELEMENT_CURRENT_EVENT', dataElementId: 'GieVkTxp4HH', programId: 'eBAyeGv0exc' }, { id: 'Zj7UnCAulEk.GieVkTxp4HH', displayName: 'Zj7UnCAulEk.GieVkTxp4HH', programRuleVariableSourceType: 'DATAELEMENT_CURRENT_EVENT', dataElementId: 'GieVkTxp4HH', programId: 'eBAyeGv0exc' }],
            id: 'lxAQ7Zs9VYR',
        },
        foundation: { programRules: [] },
        orgUnit: { id: 'DiszpKrYNg8', name: 'Ngelehun CHC' },
    },
    {
        program: {
            programRules: [{ id: 'OIaW5H2KNNp', condition: "#{Diagnosis_Method}!='Microscope'", description: 'If Diagnosis Method == Microscope or Null —> hide section “RDT                Results”', displayName: 'XX MAL RDT - DIAGNOSIS METHOD - Microscopy', programId: 'MoUd5BTQ3lY', programRuleActions: [{ id: 'JkIaHCPIYHO', programRuleActionType: 'HIDESECTION', programStageSectionId: 'PAvRSZQyYjR' }], priority: 1 }, { id: 'V2rfYrFFDAn', condition: "(#{Diagnosis_Method}=='Microscopy' && #{Diagnosis_M_Result}!= 'Pv')                || (#{Diagnosis_Method}=='RDT' && #{Diagnosis_RDT_Result} !=                'Pv')", description: 'If Diagnosis Result == Pv or Negative or Null —> hide sections Treatment                Pf, Mixed', displayName: 'XX MAL RDT - DIAGNOSIS RESULT - Pv', programId: 'MoUd5BTQ3lY', programRuleActions: [{ id: 'x8WTOVMnE0g', programRuleActionType: 'HIDESECTION', programStageSectionId: 'yEPpF2nQ2bZ' }] }, { id: 'WMsEkeKwb18', condition: "(#{Diagnosis_Method}=='Microscopy' && #{Diagnosis_M_Result}!=                'Mixed') || (#{Diagnosis_Method}=='RDT' && #{Diagnosis_RDT_Result}!=                'Mixed')", description: 'If Diagnosis Result == Mixed or Negative or Null —> hide sections Treatment                Pf, Pv', displayName: 'XX MAL RDT - DIAGNOSIS RESULT - Mixed', programId: 'MoUd5BTQ3lY', programRuleActions: [{ id: 'ZxA4tbBK7y6', programRuleActionType: 'HIDESECTION', programStageSectionId: 'OzoUDRCtylB' }], priority: 2 }, { id: 'glMKQLqV6hN', condition: "(#{Diagnosis_Method} =='Microscopy' && #{Diagnosis_M_Result} != 'Pf')                || (#{Diagnosis_Method} == 'RDT' && #{Diagnosis_RDT_Result} !=                'Pf')", description: 'f Diagnosis Result == Pf or or Negative or Null —> hide sections Treatment                Pv, Mixed', displayName: 'XX MAL RDT - DIAGNOSIS RESULT - Pf', programId: 'MoUd5BTQ3lY', programRuleActions: [{ id: 'd9dB4tQWZka', programRuleActionType: 'HIDESECTION', programStageSectionId: 'PYgwzV0eQWs' }], priority: 2 }, { id: 'tQP5ArcLjXP', condition: "#{Diagnosis_Method} !='RDT'", description: 'If Diagnosis Method == RDT or Null —> hide section                “Microscopy”', displayName: 'XX MAL RDT - DIAGNOSIS METHOD - RDT', programId: 'MoUd5BTQ3lY', programRuleActions: [{ id: 'aSKg6HVXBMP', programRuleActionType: 'HIDESECTION', programStageSectionId: 'TIpDZhpNkmS' }], priority: 1 }, { id: 'zgwtr9dSVRA', condition: "#{Diagnosis_Method} !='Not Tested'", description: 'If Diagnosis Method == Not Tested or Null —> hide section “Reason for not                testing”', displayName: 'XX MAL RDT - DIAGNOSIS METHOD - Not Tested', programId: 'MoUd5BTQ3lY', programRuleActions: [{ id: 'TNL5nLGT2Jv', programRuleActionType: 'HIDESECTION', programStageSectionId: 'YpzV7H2BA6C' }] }],
            programRuleVariables: [{ id: 'Bn9GkaU8ayh', dataElementId: 'lWLkpWMHqEq', displayName: 'Diagnosis_Method', programId: 'MoUd5BTQ3lY', programRuleVariableSourceType: 'DATAELEMENT_CURRENT_EVENT', useNameForOptionSet: false }, { id: 'Q2zYkzn2fu7', dataElementId: 'XEuy83qbOvM', displayName: 'Diagnosis_M_Result', programId: 'MoUd5BTQ3lY', programRuleVariableSourceType: 'DATAELEMENT_CURRENT_EVENT', useNameForOptionSet: false }, { id: 'i47jDXmfVOC', dataElementId: 'diH9IbKTpHj', displayName: 'Diagnosis_RDT_Result', programId: 'MoUd5BTQ3lY', programRuleVariableSourceType: 'DATAELEMENT_CURRENT_EVENT', useNameForOptionSet: false }],
            id: 'MoUd5BTQ3lY',
        },
        foundation: { programRules: [] },
        orgUnit: { id: 'DiszpKrYNg8', name: 'Ngelehun CHC' },
    },
    {
        program: null,
        foundation: null,
        orgUnit: { id: 'DiszpKrYNg8', name: 'Ngelehun CHC' },
    },
];

programs.forEach(({ program, foundation, orgUnit }) => {
    test('Tests on runRulesForSingleEvent function', () => {
        const rulesEngine = new RulesEngine(inputValueConverter, momentConverter, i18n.t, outputRulesEffectsValueConverter);

        const rulesEffects = runRulesForSingleEvent(rulesEngine, program, foundation, orgUnit);

        expect(rulesEffects).toMatchSnapshot();
    });
});
