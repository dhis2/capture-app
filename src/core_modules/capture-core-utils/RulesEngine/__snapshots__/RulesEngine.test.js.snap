// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rules engine area is hidden and given value is {"sWoqcoByYmD":false} 1`] = `
Array [
  Object {
    "id": "vANAXwtLwcT",
    "message": "Hemoglobin value lower than normal ",
    "type": "SHOWWARNING",
  },
  Object {
    "id": "Ok9OQpitjQr",
    "type": "HIDEFIELD",
  },
]
`;

exports[`rules engine area is hidden and given value is {"sWoqcoByYmD":true} 1`] = `
Array [
  Object {
    "id": "vANAXwtLwcT",
    "message": "Hemoglobin value lower than normal ",
    "type": "SHOWWARNING",
  },
]
`;

exports[`rules engine field will be hidden and given value is {"oZg33kd9taw":"Female","SWfdB5lX0fk":null} 1`] = `
Array [
  Object {
    "displayKeyValuePair": Object {
      "id": "x7PaHGvgWY2",
      "key": "BMI",
      "value": NaN,
    },
    "id": "indicators",
    "type": "DISPLAYKEYVALUEPAIR",
  },
]
`;

exports[`rules engine field will be hidden and given value is {"oZg33kd9taw":"Male","SWfdB5lX0fk":null} 1`] = `
Array [
  Object {
    "id": "SWfdB5lX0fk",
    "type": "HIDEFIELD",
  },
  Object {
    "displayKeyValuePair": Object {
      "id": "x7PaHGvgWY2",
      "key": "BMI",
      "value": NaN,
    },
    "id": "indicators",
    "type": "DISPLAYKEYVALUEPAIR",
  },
]
`;

exports[`rules engine field will be hidden and given value is {"oZg33kd9taw":null,"SWfdB5lX0fk":null} 1`] = `
Array [
  Object {
    "displayKeyValuePair": Object {
      "id": "x7PaHGvgWY2",
      "key": "BMI",
      "value": NaN,
    },
    "id": "indicators",
    "type": "DISPLAYKEYVALUEPAIR",
  },
]
`;

exports[`rules engine where BMI is calculated and given value is {"qrur9Dvnyt5":40,"GieVkTxp4HH":180,"vV9UWAZohSf":85} 1`] = `
Array [
  Object {
    "displayKeyValuePair": Object {
      "id": "x7PaHGvgWY2",
      "key": "BMI",
      "value": 26.234567901234566,
    },
    "id": "indicators",
    "type": "DISPLAYKEYVALUEPAIR",
  },
]
`;

exports[`rules engine where BMI is calculated and given value is {"qrur9Dvnyt5":40,"GieVkTxp4HH":180,"vV9UWAZohSf":null} 1`] = `
Array [
  Object {
    "displayKeyValuePair": Object {
      "id": "x7PaHGvgWY2",
      "key": "BMI",
      "value": 0,
    },
    "id": "indicators",
    "type": "DISPLAYKEYVALUEPAIR",
  },
]
`;

exports[`rules engine where BMI is calculated and given value is {"qrur9Dvnyt5":40,"GieVkTxp4HH":null,"vV9UWAZohSf":85} 1`] = `
Array [
  Object {
    "displayKeyValuePair": Object {
      "id": "x7PaHGvgWY2",
      "key": "BMI",
      "value": Infinity,
    },
    "id": "indicators",
    "type": "DISPLAYKEYVALUEPAIR",
  },
]
`;

exports[`rules engine where BMI is calculated and given value is {"qrur9Dvnyt5":40,"GieVkTxp4HH":null,"vV9UWAZohSf":null} 1`] = `
Array [
  Object {
    "displayKeyValuePair": Object {
      "id": "x7PaHGvgWY2",
      "key": "BMI",
      "value": NaN,
    },
    "id": "indicators",
    "type": "DISPLAYKEYVALUEPAIR",
  },
]
`;

exports[`rules engine where BMI is calculated and given value is {"qrur9Dvnyt5":null,"GieVkTxp4HH":180,"vV9UWAZohSf":85} 1`] = `
Array [
  Object {
    "displayKeyValuePair": Object {
      "id": "x7PaHGvgWY2",
      "key": "BMI",
      "value": 26.234567901234566,
    },
    "id": "indicators",
    "type": "DISPLAYKEYVALUEPAIR",
  },
]
`;

exports[`rules engine where BMI is calculated and given value is {"qrur9Dvnyt5":null,"GieVkTxp4HH":180,"vV9UWAZohSf":null} 1`] = `
Array [
  Object {
    "displayKeyValuePair": Object {
      "id": "x7PaHGvgWY2",
      "key": "BMI",
      "value": 0,
    },
    "id": "indicators",
    "type": "DISPLAYKEYVALUEPAIR",
  },
]
`;

exports[`rules engine where BMI is calculated and given value is {"qrur9Dvnyt5":null,"GieVkTxp4HH":null,"vV9UWAZohSf":85} 1`] = `
Array [
  Object {
    "displayKeyValuePair": Object {
      "id": "x7PaHGvgWY2",
      "key": "BMI",
      "value": Infinity,
    },
    "id": "indicators",
    "type": "DISPLAYKEYVALUEPAIR",
  },
]
`;

exports[`rules engine where BMI is calculated and given value is {"qrur9Dvnyt5":null,"GieVkTxp4HH":null,"vV9UWAZohSf":null} 1`] = `
Array [
  Object {
    "displayKeyValuePair": Object {
      "id": "x7PaHGvgWY2",
      "key": "BMI",
      "value": NaN,
    },
    "id": "indicators",
    "type": "DISPLAYKEYVALUEPAIR",
  },
]
`;

exports[`rules engine where the default values Tests on runRulesForSingleEvent function 1`] = `
Array [
  Object {
    "id": "vANAXwtLwcT",
    "message": "Hemoglobin value lower than normal ",
    "type": "SHOWWARNING",
  },
  Object {
    "id": "Ok9OQpitjQr",
    "type": "HIDEFIELD",
  },
]
`;

exports[`rules engine where the default values Tests on runRulesForSingleEvent function 2`] = `
Array [
  Object {
    "displayKeyValuePair": Object {
      "id": "x7PaHGvgWY2",
      "key": "BMI",
      "value": NaN,
    },
    "id": "indicators",
    "type": "DISPLAYKEYVALUEPAIR",
  },
]
`;

exports[`rules engine where the default values Tests on runRulesForSingleEvent function 3`] = `
Array [
  Object {
    "id": "PAvRSZQyYjR",
    "type": "HIDESECTION",
  },
  Object {
    "id": "TIpDZhpNkmS",
    "type": "HIDESECTION",
  },
  Object {
    "id": "YpzV7H2BA6C",
    "type": "HIDESECTION",
  },
]
`;

exports[`rules engine where the default values Tests on runRulesForSingleEvent function 4`] = `null`;

exports[`rules engine where value needs to >= 9 and <= 99 and given value is {"vANAXwtLwcT":0} 1`] = `
Array [
  Object {
    "id": "vANAXwtLwcT",
    "message": "Hemoglobin value lower than normal ",
    "type": "SHOWWARNING",
  },
  Object {
    "id": "Ok9OQpitjQr",
    "type": "HIDEFIELD",
  },
]
`;

exports[`rules engine where value needs to >= 9 and <= 99 and given value is {"vANAXwtLwcT":9} 1`] = `
Array [
  Object {
    "id": "Ok9OQpitjQr",
    "type": "HIDEFIELD",
  },
]
`;

exports[`rules engine where value needs to >= 9 and <= 99 and given value is {"vANAXwtLwcT":99} 1`] = `
Array [
  Object {
    "id": "Ok9OQpitjQr",
    "type": "HIDEFIELD",
  },
]
`;

exports[`rules engine where value needs to >= 9 and <= 99 and given value is {"vANAXwtLwcT":100} 1`] = `
Array [
  Object {
    "id": "vANAXwtLwcT",
    "message": "The hemoglobin value cannot be above 99 ",
    "type": "SHOWERROR",
  },
  Object {
    "id": "Ok9OQpitjQr",
    "type": "HIDEFIELD",
  },
]
`;
